### PR TITLE
[Woo POS] Improve Card Reader Payment Flow Error Handling

### LINKED ISSUE #12542: [Woo POS] Error message doesn’t feel right when failing to collect the payment due to the lack of network
❓ The error message doesn’t feel like the right one when failing to collect the payment due to the lack of network (we show error fetching order – so not sure if it’s a bad message and worth doing anything about)

https://wp.me/pdfdoF-5iv

In addition, if the card reader is not connected yet, we don't show any error message while in airplane mode.

This change will also affect existing payment flows since we're using the same dialog.

### PR DESCRIPTION
**Pull request by backwardstruck**
Closes: #12542

Branched off of this PR from @AnirudhBhat (thanks for the help):
https://github.com/woocommerce/woocommerce-android/pull/12569

### Description
Addresses an issue with error handling in the card reader payment flow. Ensures that network connectivity is checked before attempting to fetch an order.

### Steps to reproduce
0. Add products and connect reader in POS mode
1. Ensure the device is in airplane mode.
2. Attempt to collect payment.
3. Observe the error message indicating network unavailability.
4. Disable airplane mode, ensuring the device has an active internet connection.
5. Attempt to collect payment again.
6. Observe that the payment flow now proceeds as expected.

### Testing information
**Devices used:**
- Tablet emualtor

**Tested scenarios:**
- Network unavailable (airplane mode).
- Normal card reader payment flow with network available and order fetch succeeds.

### The tests that have been performed
- Verified normal payment flow with successful order fetch and payment collection.
- Manual smoke tests on both tablet and phone devices to ensure no regressions were introduced.

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.  
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.  
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by backwardstruck** [comment]
@AnirudhBhat this is still in draft because I want to make sure I have the right approach. What do you think about me checking for connection before attempting to fetch the order? 

Should I instead watch the connection state throughout the payment process or try to return an error from the order detail repository?

**Comment by AnirudhBhat** [comment]
Thanks for working on this improvement @backwardstruck!

>this is still in draft because I want to make sure I have the right approach. What do you think about me checking for connection before attempting to fetch the order?
Should I instead watch the connection state throughout the payment process or try to return an error from the order detail repository?

The approach sounds fine. However, as I understand it, one of the key principles we’ve tried to maintain since the inception of POS is to minimise changes to the existing IPP store management code in order to avoid potential regressions. With that in mind, we could consider the following alternative:

In `WooPosTotalsViewModel`, we can check the internet connection before initiating the payment process, like this:

```
private fun collectPayment() {
    if (!wooPosNetworkStatus.isConnected()) {
        viewModelScope.launch {
            childrenToParentEventSender.sendToParent(ChildToParentEvent.NoInternet) // This event is being handled in the WooPosHomeViewModel - PR 12569
        }
    } else {
        val orderId = dataState.value.orderId
        check(orderId != EMPTY_ORDER_ID)
        cardReaderFacade.collectPayment(orderId)
    }
}
```
I implemented a similar solution while resolving https://github.com/woocommerce/woocommerce-android/issues/12497, and you can refer to the relevant code here: https://github.com/woocommerce/woocommerce-android/pull/12569. You can merge this changes into your branch and start modifying if you agree with this approach.

This approach allows us to resolve the issue without modifying the existing IPP-related code.

That said, I agree that in the current IPP store management solution, when there is no internet and we try to collect a payment, the "error fetching order" dialog is shown instead of a "No internet" message. However, in store management mode, a persistent offline banner is displayed at the bottom whenever we lose connectivity, giving users an indication of the issue. While I agree that the UI could be improved here, I think that can be addressed in a separate PR.

Screenshot 2024-09-10 at 10 24 19 AM

Let me know your thoughts on this approach!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt [hunk 1]
```diff
@@ -80,9 +82,15 @@ class WooPosTotalsViewModel @Inject constructor(
...
-        check(orderId != EMPTY_ORDER_ID)
-        cardReaderFacade.collectPayment(orderId)
+        if (!networkStatus.isConnected()) {
+            viewModelScope.launch {
+                childrenToParentEventSender.sendToParent(ChildToParentEvent.NoInternet)
+            }
+        } else {
+            val orderId = dataState.value.orderId
+            check(orderId != EMPTY_ORDER_ID)
+            cardReaderFacade.collectPayment(orderId)
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:92]: ❓ What do you think about adding test to verify that `cardReaderFacade.collectPayment` is not called when there is no internet?

**Review by AnirudhBhat — APPROVED** [review]
Thanks for working on this @backwardstruck. LGTM! I've added a minor comment regarding the inclusion of a unit test. I'm holding off on merging this since it's based on another PR; we'll proceed once that has been merged.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt (hunk 1 above)
Comment (backwardstruck) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:92]: Yes, you had suggested that earlier but I failed to add that test.

Comment (backwardstruck) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:92]: Added.

**Pull request merged**